### PR TITLE
Add monospace toggle button to textarea

### PR DESCRIPTION
**Pull request by silverwind**
- Add new button to textarea to switch font. State is persisted in localStorage.
- Change markdown-switch-easymde button from `<span>` to `<button>`
- Slightly increased monospace font globally by 5% as I think it fits better.

For hover effect on these buttons I'm deferring to https://github.com/go-gitea/gitea/pull/23896.

### CODE REVIEW

**Comment by silverwind** [comment]
I guess I would prefer if we could use the server-side user settings storage for settings like these instead of volatile localStorage. Is there API available? If not, I guess we defer to migrating these localStorage settings there later.

**Comment by jolheiser** [comment]
> I guess I would prefer if we could use the server-side user settings storage for settings like these instead of potentially volatile localStorage. Is there API available? If not, I guess we defer to migrating these localStorage settings there later.

There is https://github.com/go-gitea/gitea/blob/main/models/user/setting.go, but you would have to wire it up to that still.

**Comment by silverwind** [comment]
> > I guess I would prefer if we could use the server-side user settings storage for settings like these instead of potentially volatile localStorage. Is there API available? If not, I guess we defer to migrating these localStorage settings there later.
> 
> ~There is https://github.com/go-gitea/gitea/blob/main/models/user/setting.go, but you would have to wire it up to that still.~ I misread, you were asking about the Gitea API I think, in which case I don't believe there's an API for the K/V table afaik.

Should be an API to store per-user settings as K/V with value being arbitrary JSON. I guess I will defer that effort to later, but I do plan opening an issue describing what I want.

File: templates/shared/combomarkdowneditor.tmpl
Comment (delvh) [inline templates/shared/combomarkdowneditor.tmpl]: Why an inline script?
Why not add another `init` method to `index.js`?

File: web_src/css/editor-markdown.css
Comment (delvh) [inline web_src/css/editor-markdown.css]: Ah yes, the conflict with #23896 you mentioned.

File: web_src/js/features/comp/ComboMarkdownEditor.js
Comment (delvh) [inline web_src/js/features/comp/ComboMarkdownEditor.js]:
```suggestion
      this.textarea.classList.toggle('gt-mono', enabled);
```

File: templates/shared/combomarkdowneditor.tmpl
Comment (delvh) [reply templates/shared/combomarkdowneditor.tmpl]: Which you already seem to be doing below?
Now I'm especially confused…

**Comment by delvh** [comment]
By the way, what's the benefit of adding this setting?
Why would you need to choose between mono-space fonts and non-mono-space fonts?

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Inline script is required to synchronously set correct textarea class for the font, otherwise it will flash wrong font on page load until `index.js` is parsed and executed.

This is very similar to the file tree on diff, we can't delay setting the state, otherwise there will be wrong flashes.

There was also https://github.com/go-gitea/gitea/pull/21429 which achieves the same without inline scripts, but I couldn't get acceptance for the method used there. Maybe it should be revisited.

File: web_src/css/editor-markdown.css
Comment (silverwind) [reply web_src/css/editor-markdown.css]: Will remove this when https://github.com/go-gitea/gitea/pull/23896 lands hopefully before this one.

File: web_src/js/features/comp/ComboMarkdownEditor.js
Comment (silverwind) [reply web_src/js/features/comp/ComboMarkdownEditor.js]: Seems equivalent. It seems so odd that this argument is called `force`, when it should be really called `present` or something.

> If set to false, then token will only be removed, but not added. If set to true, then token will only be added, but not removed.

**Comment by silverwind** [comment]
Monospace is nice when you write code and want to align the indendation properly. Some people even prefer monospace all the time. GitHub also has a setting for it at the bottom of https://github.com/settings/appearance.

**Comment by delvh** [comment]
Blocked by #23896.

**Comment by yardenshoham** [comment]
https://github.com/go-gitea/gitea/pull/23896 was merged

**Review by HesterG — APPROVED** [review]

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [inline templates/shared/combomarkdowneditor.tmpl]: I guess it still needs longEnoughName to prevent from global name conflicting.

Or, this script could be `type=module`

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [inline templates/shared/combomarkdowneditor.tmpl]: These lines could be removed, and let `ComboMarkdownEditor.js` do so, to keep the inline script code as less as possible.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Yes, could give even longer name. Can't be `module` as it would load too late. Module scripts are automatically deferred.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Yeah, the button attributes have no impact on rendering, so could be removed here.

File: web_src/css/editor-markdown.css
Comment (silverwind) [reply web_src/css/editor-markdown.css]: Done, color will stay as the native `<button>` elements need it to prevent browser UA style to color them.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Variable is completely removed now.

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: Inline script is unavoidable, otherwise there is a flash of wrong content.

File: templates/shared/combomarkdowneditor.tmpl
Comment (wxiaoguang) [inline templates/shared/combomarkdowneditor.tmpl]: Generally LGTM

A nit, I think the key could be `markdown-editor-monospace`, because there are not only one editor in Gitea (eg: there is also a code editor).

**Review by lunny — APPROVED** [review]

File: templates/shared/combomarkdowneditor.tmpl
Comment (silverwind) [reply templates/shared/combomarkdowneditor.tmpl]: done

**Comment by silverwind** [comment]
.

**Pull request merged**
